### ISSUE TITLE
Add default selector to all components

### ISSUE THREAD

**Issue by simonihmig** [simplabs/ember-test-selectors#62]
Before having the ability to just add data-test attributes to any component (available since 0.1.0), I was using a Mixin like the following which I manually added to each and every component:

```js
Ember.Mixin.create({
  attributeBindings: 'data-test-selector',

  'data-test-selector': computed(function() {
    return Object.getPrototypeOf(this)._debugContainerKey.replace(/^component:/, '');
  })
});
```

The new version of this addon is truly great in that I can assign data-test attributes everywhere and they just work, and everything is cleanly wiped out for production builds. But the missing piece is still having a default attribute, so I have not to assign them for each and every component.

I removed the afore mentioned mixin and replaced it with an vendor-import approach that you also used, which works and would be fine for me. 

But I wonder if it would make sense to add that by default, so this comes for free for every user!?

For example this comes handy when testing components with `ember-cli-page-object`, so every page-object-components `.scope` (http://ember-cli-page-object.js.org/docs/v1.8.x/components#scopes) can just be set to the default test-selector for that component.

Should be easy to implement, I can pull a PR together if you guys agree on this!

**Comment by marcoow** [comment]
I initially though this was a good idea as well but @bcardarella and @lukemelia convinced me it's not so we have decided we won't add that - see #26 for a lengthy discussion about the topic.

**Comment by simonihmig** [comment]
@marcoow Thanks for pointing this out, did not know this has been discussed before.

While understanding the argument, I think there are other valid ones as well. Most importantly a pragmatic one, as I don't want to manually assign selectors in an app with something like a hundred components. And if I would have to, I would certainly just use the same names as the component, what reason would I have to assign a different name? So no difference...

For the renaming use case, would not be a big deal, if using `ember-cli-page-object` it's one line in one file. If not you could still just override the default selector to make it the same as it was before. 

So far have not renamed a single component in this current app, but definitely would have to assign >100 component selectors... 

So I would come to a different conclusion. But that's ok, no big deal. Just hope that other (newbie) users are ok with this as well...

**Comment by marcoow** [comment]
> While understanding the argument, I think there are other valid ones as well. Most importantly a pragmatic one, as I don't want to manually assign selectors in an app with something like a hundred components. And if I would have to, I would certainly just use the same names as the component, what reason would I have to assign a different name? So no difference...

The reasoning behind this decision is that it would encourage brittle tests. Actually when looking at the DOM from the perspective of an acceptance or integration test you shouldn't care what component renders a particular element but instead you care about what a particular element _"means"_ in the context of the text. Also the same component might mean different things in different contexts (think of e.g. a `{{confirm-button}}` component that could be used for deleting or blocking a user etc.). Tests should not be coupled to the use of concrete components as they should also not break when e.g. the component is renamed or replaced with a different but the tested functionality is still intact.

Of course you can always define your own mixin that assigns a data attribute that equals the component's name.

**Comment by Turbo87** [comment]
@marcoow @simonihmig as I mentioned a few days ago I'm on the fence here. I tend to agree that during refactoring this might cause more work, but then again you have a lot more work up front now to actually setup the selectors.

I personally would be okay with implementing this, as long as it's possible to disable it in the plugin config, but I think @marcoow has pretty strong feelings about this 😉

**Comment by marcoow** [comment]
Yeah, I initially had pretty strong feelings supporting this but was convinced I should have pretty strong feelings against it actually ;)

Renaming the component isn't really the only problem. You might also refactors whole component structures or move from one addon to another so that the component name changes because of that or even replace the component with a native element (think of the above example where the `{{confirm-button}}` might be replaced with a regular button). In all these scenarios the functionality under test might stay intact perfectly well and the test shouldn't break.

**Comment by Turbo87** [comment]
> I tend to agree that during refactoring this might cause more work, but then again you have a lot more work up front now to actually setup the selectors.

😉 

I'm not quite convinced yet that the work during refactoring is that much larger than the work for the initial setup.

Since many people are asking for this there is certainly a demand for such a feature and that is mainly the reason why I think it would make sense to offer this as an **option**. We could still warn about the implications on using it in the README.

**Comment by marcoow** [comment]
I'd say it's easy enough to just add an initializer or so that does it per app actually. I think it's important to not support something that would be considered a bad practice (for the reasons given above) in an addon like this.

**Comment by lukemelia** [comment]
I don't want to have the same conversation as #26 again, but I'll just add:

1) there is not really any such thing as "initial setup". In practice, you add data-test selectors as you write tests
2) The entire purpose of using this addon is to keep your acceptance tests stable in the face of refactorings that don't actually affect functionality. The proposed feature would create [Connascence of Name](https://en.wikipedia.org/wiki/Connascence_(computer_programming)#Connascence_of_Name_.28CoN.29), that is changing the name of components (which don't affect the end user) would cause you to have to update acceptance tests despite the fact that nothing has changed for the user.

**Comment by simonihmig** [comment]
> there is not really any such thing as "initial setup". In practice, you add data-test selectors as you write tests

Well, maybe we have used this in a different (wrong?) way, but actually we did provide each component a test selector upfront (through the use of the mentioned mixin). For low level things like a button we certainly overrode this is certain places to distinguish them by their function (save, cancel, whatever). But for more high level components we were actually sticking to the default selector and had no need to customize it.

Also as I mentioned, we used this in conjunction with `ember-cli-page-object`, that gives you an extra (somewhat different) level of abstraction and decoupling from your tests, so having to change selectors during refactorings was no real concern to us. (as this would happen in just one place)

Anyways, as I said I can live with it. Would still support an option as @Turbo87 suggested, but that decision is now yours! :)
